packages/backports/installer/cos.sh

```diff
@@ -0,0 +1,1367 @@
+        source /etc/os-release
+    fi
+
+    if [ -e /etc/cos/config ]; then
+        source /etc/cos/config
+    fi
+}
+
+prepare_chroot() {
+    local dir=$1
+
+    for mnt in /dev /dev/pts /proc /sys
+    do
+        mount -o bind $mnt $dir/$mnt
+    done
+}
+
+cleanup_chroot() {
+    local dir=$1
+
+    for mnt in /sys /proc /dev/pts /dev
+    do
+        umount $dir/$mnt
+    done
+}
+
+run_hook() {
+    local hook=$1
+    local dir=$2
+
+    prepare_chroot $dir
+    chroot $dir /usr/bin/cos-setup $hook
+    cleanup_chroot $dir
+}
+
+
+is_mounted() {
+    mountpoint -q "$1"
+}
+
+is_booting_from_squashfs() {
+    if cat /proc/cmdline | grep -q "COS_RECOVERY"; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+prepare_deploy_target() {
+    if [ -f  ${STATEDIR}/cOS/active.img ] && [ "$FORCE" != "true" ]; then
+        echo "There is already an active deployment in the system, use '--force' flag to overwrite it"
+        exit 1
+    fi
+
+    mkdir -p ${STATEDIR}/cOS || true
+    rm -rf ${STATEDIR}/cOS/active.img || true
+    dd if=/dev/zero of=${STATEDIR}/cOS/active.img bs=1M count=$DEFAULT_IMAGE_SIZE
+    mkfs.ext2 ${STATEDIR}/cOS/active.img
+    mount -t ext2 -o loop ${STATEDIR}/cOS/active.img $TARGET
+}
+
+prepare_target() {
+    mkdir -p ${STATEDIR}/cOS || true
+    rm -rf ${STATEDIR}/cOS/transition.img || true
+    dd if=/dev/zero of=${STATEDIR}/cOS/transition.img bs=1M count=$DEFAULT_IMAGE_SIZE
+    mkfs.ext2 ${STATEDIR}/cOS/transition.img
+    mount -t ext2 -o loop ${STATEDIR}/cOS/transition.img $TARGET
+}
+
+usage()
+{
+    echo "Usage: $PROG install|deploy|upgrade|reset|rebrand [options]"
+    echo ""
+    echo "Example: $PROG-install /dev/vda"
+    echo "  install:"
+    echo "  [--partition-layout /path/to/config/file.yaml ] [--force-efi] [--force-gpt] [--iso https://.../OS.iso] [--debug] [--tty TTY] [--poweroff] [--no-format] [--config https://.../config.yaml] DEVICE"
+    echo ""
+    echo "  upgrade:"
+    echo "  [--strict] [--recovery] [--no-verify] [--no-cosign] [--directory] [--docker-image] (IMAGE/DIRECTORY)"
+    echo ""
+    echo "  deploy:"
+    echo "  [--strict] [--no-verify] [--no-cosign] [--force] [--docker-image] (IMAGE)"
+    echo ""
+    echo "   DEVICE must be the disk that will be partitioned (/dev/vda). If you are using --no-format it should be the device of the COS_STATE partition (/dev/vda2)"
+    echo "   IMAGE must be a container image if --docker-image is specified"
+    echo "   DIRECTORY must be passed if --directory is specified"
+    echo ""
+    echo "The parameters names refer to the same names used in the cmdline, refer to README.md for"
+    echo "more info."
+    echo ""
+    exit 1
+}
+
+## END COMMON
+
+## INSTALLER
+umount_target() {
+    sync
+    umount ${TARGET}/oem
+    umount ${TARGET}/usr/local
+    umount ${TARGET}/boot/efi || true
+    umount ${TARGET}
+    if [ -n "$LOOP" ]; then
+        losetup -d $LOOP
+    fi
+}
+
+installer_cleanup2()
```

Comment:
can we have a bit more descriptive name here?

Reply:
that's not really new code here, I'd try to keep the changes stick to the patch backport, or we might introduce new bugs

Reply:
here is the real diff between the two `cos.sh`:

```
--- packages/installer/cos.sh   2021-12-07 15:27:19.112633623 +0100
+++ cos.sh      2021-12-07 15:29:08.209264737 +0100
@@ -198,6 +198,12 @@ prepare_passive() {
 
 part_probe() {
     local dev=$1
+
+    # Don't require udevadm necessarly, but run it best-effort
+    if hash udevadm 2>/dev/null; then
+        udevadm settle
+    fi
+
     partprobe ${dev} 2>/dev/null || true
 
     sync
@@ -240,8 +246,11 @@ do_format()
         if [ "$BOOTFLAG" == "esp" ]; then
             parted -s ${DEVICE} mkpart primary fat32 0% 50MB # efi
             parted -s ${DEVICE} set 1 ${BOOTFLAG} on
+
+            part_probe $DEVICE
+
             PREFIX=${DEVICE}
-            if [ ! -e ${PREFIX}${STATE_NUM} ]; then
+            if [ ! -e ${PREFIX}1 ]; then
                 PREFIX=${DEVICE}p
             fi
             BOOT=${PREFIX}1
@@ -250,6 +259,7 @@ do_format()
         elif [ "$BOOTFLAG" == "bios_grub" ]; then
             parted -s ${DEVICE} mkpart primary 0% 1MB # BIOS boot partition for GRUB
             parted -s ${DEVICE} set 1 ${BOOTFLAG} on
+            part_probe $DEVICE
         fi
 
         yip -s partitioning $COS_PARTITION_LAYOUT
```